From 99646423a87cb6c7a8e42def5b5db657939f7109 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:09:02 -0400
Subject: [powerpc] Revert c6102609 and replace it with the correct fix for vio dma mask setting

Message-id: <1410545655-205645-314-git-send-email-dzickus@redhat.com>
Patchwork-id: 94432
O-Subject: [RHEL7 PATCH 313/626] powerpc: Revert c6102609 and replace it with the correct fix for vio dma mask setting
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 10862a0c71bff5c7b88812d6a9a9c0be56bc659a
Author: Li Zhong <zhong@linux.vnet.ibm.com>
Date:   Thu Nov 28 17:22:25 2013 +0800

    powerpc: Revert c6102609 and replace it with the correct fix for vio dma mask setting

    This patch reverts my previous "fix", and replace it with the correct
    fix from Russell.

    And as Russell pointed out -- dma_set_mask_and_coherent() (and the other
    dma_set_mask() functions) are really supposed to be used by drivers
    only.

    Signed-off-by: Li Zhong <zhong@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/vio.c b/arch/powerpc/kernel/vio.c
index 2233166..93a5922 100644
--- a/arch/powerpc/kernel/vio.c
+++ b/arch/powerpc/kernel/vio.c
@@ -1426,7 +1426,8 @@ struct vio_dev *vio_register_device_node(struct device_node *of_node)
 
   /* needed to ensure proper operation of coherent allocations
    * later, in case driver doesn't set it explicitly */
-  dma_coerce_mask_and_coherent(&viodev->dev, DMA_BIT_MASK(64));
+  viodev->dev.coherent_dma_mask = DMA_BIT_MASK(64);
+  viodev->dev.dma_mask = &viodev->dev.coherent_dma_mask;
  }
 
  /* register with generic device framework */
-- 
1.7.1